frontend/src/metabase/public/containers/PublicAction/PublicActionLoader.tsx

```diff
@@ -0,0 +1,73 @@
+import React, { useCallback, useState } from "react";
+import { connect } from "react-redux";
+
+import { useOnMount } from "metabase/hooks/use-on-mount";
+import { useSafeAsyncFunction } from "metabase/hooks/use-safe-async-function";
+import { setErrorPage } from "metabase/redux/app";
+import { PublicApi } from "metabase/services";
+
+import type { WritebackAction } from "metabase-types/api";
+import type { AppErrorDescriptor } from "metabase-types/store";
+
+import EmbedFrame from "../../components/EmbedFrame";
+import PublicAction from "./PublicAction";
+import {
+  LoadingAndErrorWrapper,
+  ContentContainer,
+} from "./PublicAction.styled";
+
+interface OwnProps {
+  params: { uuid: string };
+}
+
+interface DispatchProps {
+  setErrorPage: (error: AppErrorDescriptor) => void;
+}
+
+type Props = OwnProps & DispatchProps;
+
+const mapDispatchToProps = {
+  setErrorPage,
+};
+
+function PublicActionLoader({ params, setErrorPage }: Props) {
+  const [action, setAction] = useState<WritebackAction | null>(null);
+  const fetchAction = useSafeAsyncFunction(PublicApi.action);
+
+  useOnMount(() => {
+    async function loadAction() {
+      try {
+        const action = await fetchAction({ uuid: params.uuid });
+        setAction(action);
+      } catch (error) {
+        setErrorPage(error as AppErrorDescriptor);
+      }
+    }
+    loadAction();
+  });
+
+  const renderContent = useCallback(() => {
+    if (!action) {
+      return null;
+    }
+    return (
+      <ContentContainer>
+        <PublicAction
+          action={action}
+          publicId={params.uuid}
+          onError={setErrorPage}
+        />
+      </ContentContainer>
+    );
+  }, [action, params.uuid, setErrorPage]);
+
+  return (
+    <EmbedFrame footerVariant="big">
+      <LoadingAndErrorWrapper loading={!action}>
+        {renderContent}
```

Comment:
Whoa! Didn't realize we can pass a render function to `<LoadingAndErrorWrapper />`

https://github.com/metabase/metabase/blob/644011426b8bc0fbf3c6092ba507ed249b0072be/frontend/src/metabase/components/LoadingAndErrorWrapper.jsx#L94-L95

Reply:
Yup!